ports: add azure_sdk port documentation (#124)
phoenix-rtos/phoenix-rtos-doc: ports/azure_sdk.md

  - [Network setup on ia32-generic-qemu](#network-setup-on-ia32-generic-qemu)
  - [Running the IoThub Client sample](#running-the-iothub-client-sample)
- [Using azure-iot-sdk-c](#using-azure-iot-sdk-c)
- [Running tests](#running-tests)
- [Known bugs](#known-bugs)

## General information

There are stored adaptations needed to run `azure-iot-sdk-c` on Phoenix-RTOS.

Azure IoT C Software Development Kit provides the interface to communicate easily with Azure IoT Hub, Azure IoT Central, and to Azure IoT Device Provisioning. It's intended for apps written in C99 or later or C++. For more information please visit the [Azure IoT C SDK Github](https://github.com/Azure/azure-iot-sdk-c).

## Supported version

The supported version is [LTS_01_2022](https://github.com/Azure/azure-iot-sdk-c/tree/LTS_01_2022_Ref01).

## Quickstart for ia32-generic-qemu

This simple guide is recommended for getting familiar with using `azure_sdk` on Phoenix-RTOS. You will need to have a `hub` created in [Azure IoT Hub](https://azure.microsoft.com/en-us/services/iot-hub/#overview) and some device within it. Each device has its connection string, which will be needed in the following instructions.

> First, please build the project with the `ports` component and set `AZURE_CONNECTION_STRING` (see above):

Author: Yeah, that's right

  ```
  echo "allow virbr0" > /etc/qemu/bridge.conf
  sudo chmod a+rw /etc/qemu/bridge.conf
  ```

  - The `/etc/qemu` directory may not exist, then create it and provide the required privileges:

  ```
  sudo mkdir /etc/qemu
  sudo chmod a+rw /etc/qemu
  echo "allow virbr0" > /etc/qemu/bridge.conf
  sudo chmod a+rw /etc/qemu/bridge.conf
  ```

  <img src="_images/azure_sdk_vibr_setup.png" width="700px">

  - Sources: https://apiraino.github.io/qemu-bridge-networking/, https://mike42.me/blog/2019-08-how-to-use-the-qemu-bridge-helper-on-debian-10

 2. Launch `qemu` using a starting script with `net` suffix:

  ```
  ./scripts/ia32-generic-qemu-net.sh
  ```

 3. Configure network and run `ash` (Busybox applet) using `rc` script:

  - Note: `IP` is assigned using `DHCP`, there are also other possibilities, for more please check the configuration file located in `_projects/ia32-generic-qemu/rootfs-overlay/etc/rc.conf.d/network`

  - Note: There are also executed other programs, for more information please check the content of the `rc` file for `ia32-generic-qemu` in `_projects/ia32-generic-qemu/rootfs-overlay/etc/rc`

  ```
  /linuxrc
  ```

>   - As You can see, the advanced version of `Phoenix-RTOS` with `POSIX` shell has been started:

Reviewer: ```suggestion
  - As you can see, the advanced version of `Phoenix-RTOS` with `POSIX` shell has been started:
```
I also don't know, what you mean by `advanced`. You might have meant `expanded` or `extended`?

  <img src="_images/azure_linuxrc.png" width="700px">

>   - Now You can check the internet connection using the `ping` applet.

Reviewer: ```suggestion
  - Now you can check the internet connection using `ping` applet.
```

>   - To make the configuration complete please set the current date using the `date` applet, like below:

Reviewer: ```suggestion
  - To complete the configuration please set the current date using `date` applet, like below:
```

  <img src="_images/azure_date.png" width="700px">

>   - Note: The required argument entered after `@` is EPOCH (POSIX time format) - You can gt it for example from [here](https://www.epochconverter.com/).

Reviewer: ```suggestion
  - Note: The required argument entered after `@` is EPOCH (POSIX time format) - you can get it for example from [here](https://www.epochconverter.com/).
```

### Running the IoThub Client sample

> Providing the internet connection is working and the current date is set, You can run the built `iothub_ll_telemetry_sample`:

Reviewer: ```suggestion
Provided that internet connection is working and the current date is set, you can run the build `iothub_ll_telemetry_sample`:
```

Author: I meant sample, which have been built. However, removing it seems to be a better idea.

```
/bin/iothub_ll_telemetry_sample
```

<img src="_images/azure_sample.png" width="700px">

>  - Note: The entered connection string was added to the sample's source code, You can find it in `_build/ia32-generic-qemu/azure_sdk/azure-iot-sdk-c-lts_01_2022/iothub_client/samples/iothub_ll_telemetry_sample/iothub_ll_telemetry_sample.c`.

Reviewer: ```suggestion
 - Note: The entered connection string was added to the sample's source code. You can find it in `_build/ia32-generic-qemu/azure_sdk/azure-iot-sdk-c-lts_01_2022/iothub_client/samples/iothub_ll_telemetry_sample/iothub_ll_telemetry_sample.c`.
```

> You can read the sent messages from Azure, for example using `AzureIotHub VS Code Extension` by clicking `Start Monitoring Built-in Event Endpoint`:

Reviewer: ```suggestion
You can read messages received from Azure, for example using `AzureIotHub VS Code Extension` by clicking `Start Monitoring Built-in Event Endpoint`:
```

 ```
[IoTHubMonitor] [6:07:21 PM] Message received from [ia32-generic-qemu]:
"test_message"
[IoTHubMonitor] [6:07:21 PM] Message received from [ia32-generic-qemu]:
"test_message"
[IoTHubMonitor] [6:07:21 PM] Message received from [ia32-generic-qemu]:
"test_message"
[IoTHubMonitor] [6:07:21 PM] Message received from [ia32-generic-qemu]:
"test_message"
[IoTHubMonitor] [6:07:21 PM] Message received from [ia32-generic-qemu]:
"test_message"
```

## Using azure-iot-sdk-c

> The above guide only shows how to run one of the provided samples, if You want to write Your own programs using the sdk please follow the next instructions. It may be helpful for the other architectures, like `armv7m7-imxrt106x-evk`, where the generated sample may not work. That's the reason why the following example is adjusted to the configuration with `mbedtls` intended for 'smaller' targets (now only the `imxrt106x` is supported). If You want to write Your own programs intented for `openssl` configuration ('larger' targets, like `ia32-generic-qemu`) there will be a few differences, like no need for setting the certificate in code and increasing stack size in Makefile.

Reviewer: ```suggestion
The above guide shows how to run only one of the provided samples. To write your own programs using the SDK, please read the following instructions. It may be helpful for other architectures, like `armv7m7-imxrt106x-evk`, where the previously generated sample may not work. That's the reason, why the following example is adjusted to the configuration with `mbedtls` intended for 'smaller' targets (now only the `imxrt106x` is supported). If you want to write your own programs intended for the `openssl` configuration ('larger' targets, like `ia32-generic-qemu`) there are few differences, e.g. there is no need for setting the certificate in code and increasing the stack size in Makefile.
```

> To use functions provided by `azure-iot-sdk-c` please place the specific `azure` libraries (and possibly `mbedtls` libraries) in `LIBS` variable in `Makefile` and include the required header file. Below is the example of an elementary program similar to `iothub_client_ll_telemetry_sample`, You can place it for example in the `_user` directory:

Reviewer: ```suggestion
To use functions provided by `azure-iot-sdk-c` please place append specific `azure` libraries (and possibly `mbedtls` libraries) to `LIBS` variable in `Makefile` and include the required header file. Below is an example of a simple program similar to `iothub_client_ll_telemetry_sample`. You can place it for example in the `_user` directory:
```

>  - Note: The SDK was tested for the `armv7m7-imxrt106x` target architecture with the `mbedtls` port, but remember that internet connection and current date have to be also provided.

Reviewer: ```suggestion
 - Note: The SDK was tested for the `armv7m7-imxrt106x` target architecture with the `mbedtls` port, but remember that internet connection and the current date also have to be provided.
```

>  - Makefile - linking with azure SDK and mbedtls libraries.

Reviewer: ```suggestion
 - Makefile - linking with Azure SDK and mbedtls libraries.
```
I don't know, how `Azure` should be spelled - starting from a capital letter or a small letter, but please be consistent with it.

    /* Setting connection status callback to get indication of connection to iothub */
    IoTHubDeviceClient_LL_SetConnectionStatusCallback(*devhandle, azure_connectionStatusCallback, NULL);
    return 0;
  }

  static void azure_close(IOTHUB_DEVICE_CLIENT_LL_HANDLE *devhandle)
  {
    IoTHubDeviceClient_LL_Destroy(*devhandle);
  }

  static void azure_deinit(void)
  {
    IoTHub_Deinit();
  }

  static void azure_init(void)
  {
    IoTHub_Init();
  }

  int main(void)
  {
      IOTHUB_DEVICE_CLIENT_LL_HANDLE devhandle;

      azure_init();
      azure_open(AZURE_CONNECTION_STRING, cert, &devhandle);

      azure_sendMsg(&devhandle, "test message");

      azure_close(&devhandle);
      azure_deinit();

      return 0;
  }
  ```

- Note: Please remember that `PORTS_AZURE_SDK` and environment variables for other required ports should be set to `y` in the specific building script in `_projects` directory or using an environment variable.

## Running tests

To build `azure_sdk` tests please set `LONG_TEST=y` environment variable before calling `build.sh`.

In the result unit tests for the `c-utility` component should be placed in the `/bin` directory.

> The tests have `ut_exe` suffix, for example: `connectionstringparser_ut_exe`, You can run it as follows:

Reviewer: ```suggestion
The tests have `ut_exe` suffix, for example: `connectionstringparser_ut_exe`. You run it as follows:
```
